File: src/fuser/memory_base/writer.py
import json
import logging
from datetime import datetime
from pathlib import Path
from typing import Optional

from fuser.knowledge_base.base_retriever import Document
from providers.singleton import singleton


@singleton
class MemoryWriter:
    """Write interactions and summaries to daily markdown files.

    Parameters
    ----------
    memory_root : str or Path, optional
        Root directory for memory storage.
    """

    def __init__(
        self,
        memory_root: str | Path | None = None,
    ):
        if memory_root is None:
            project_root = Path(__file__).parent.parent.parent.parent
            memory_root = project_root / "memory"
        self.memory_root = Path(memory_root)
        self.daily_dir = self.memory_root / "daily"
        self.users_dir = self.memory_root / "users"
        self._ensure_dirs()

    def _ensure_dirs(self) -> None:
        """Create memory directories if they don't exist."""
        self.daily_dir.mkdir(parents=True, exist_ok=True)
        self.users_dir.mkdir(parents=True, exist_ok=True)

    def _get_daily_path(self) -> Path:
        """Get the path for today's daily log file."""
        today = datetime.now().strftime("%Y-%m-%d")
        return self.daily_dir / f"{today}.md"

    def ensure_user_dir(self, user_id: str) -> Path:
        """Create the user directory and initialize profile/facts if needed.

        Parameters
        ----------
        user_id : str
            Normalized user identity.

        Returns
        -------
        Path
            Path to the user directory.
        """
        user_dir = self.users_dir / user_id
        user_dir.mkdir(parents=True, exist_ok=True)

        profile_path = user_dir / "profile.json"
        if not profile_path.exists():
            now = datetime.now().isoformat(timespec="seconds")
            profile = {
                "user_id": user_id,
                "display_name": user_id.title(),
                "first_seen": now,
                "last_seen": now,
                "interaction_count": 0,
            }
            profile_path.write_text(json.dumps(profile, indent=2), encoding="utf-8")
            logging.info(f"Memory: created user profile for {user_id}")

        facts_path = user_dir / "facts.json"
        if not facts_path.exists():
            facts = {"user_id": user_id, "facts": []}
            facts_path.write_text(json.dumps(facts, indent=2), encoding="utf-8")

        return user_dir

    def update_user_profile(self, user_id: str) -> None:
        """Update last_seen and increment interaction_count."""
        profile_path = self.users_dir / user_id / "profile.json"
        if not profile_path.exists():
            return

        try:
            profile = json.loads(profile_path.read_text(encoding="utf-8"))
            profile["last_seen"] = datetime.now().isoformat(timespec="seconds")
            profile["interaction_count"] = profile.get("interaction_count", 0) + 1
            profile_path.write_text(json.dumps(profile, indent=2), encoding="utf-8")
        except Exception as e:
            logging.warning(f"Memory: failed to update profile for {user_id}: {e}")

    def append_interaction(
        self,
        user_msg: str,
        actions: list,
        user_id: Optional[str] = None,
    ) -> None:
        """Append a user-robot interaction to today's daily log.

        Parameters
        ----------
        user_msg : str
            ASR input.
        actions : list
            LLM output actions.
        user_id : str, optional
            Identity of the user (from face recognition).
        """
        if not user_msg.strip():
            return

        if user_id:
            self.ensure_user_dir(user_id)
            self.update_user_profile(user_id)

        daily_path = self._get_daily_path()
        timestamp = datetime.now().strftime("%H:%M:%S")

        robot_msg = " | ".join(f"{a.type}: {a.value}" for a in actions if a.value)

        entry = f"\n## {timestamp}\n"
        if user_id:
            entry += f"[User: {user_id}]\n"
        entry += f"- **User**: {user_msg.strip()}\n"
        entry += f"- **Robot**: {robot_msg}\n"

        try:
            with open(daily_path, "a", encoding="utf-8") as f:
                f.write(entry)
            logging.debug(f"Memory: appended interaction to {daily_path.name}")
        except Exception as e:
            logging.error(f"Memory: failed to write interaction: {e}")

    async def append_to_index(
        self,
        user_msg: str,
        actions: list,
        user_id: Optional[str] = None,
    ) -> None:
        """Embed and insert a new interaction into the in-memory index.

        Skipped if the index has not been initialized yet.

        Parameters
        ----------
        user_msg : str
            ASR input.
        actions : list
            LLM output actions.
        user_id : str, optional
            Identity of the user (from face recognition).
        """
        if not user_msg.strip():
            return

        from fuser.memory_base.reader import MemoryReader

        reader = MemoryReader()
        if not reader._index_initialized:
            return

        try:
            date_str = datetime.now().strftime("%Y-%m-%d")
            timestamp = datetime.now().strftime("%H:%M:%S")
            robot_msg = " | ".join(f"{a.type}: {a.value}" for a in actions if a.value)

            user_tag = f"[User: {user_id}]\n" if user_id else ""
            text = (
                f"[Date: {date_str}]\n## {timestamp}\n{user_tag}"
                f"- **User**: {user_msg.strip()}\n"
                f"- **Robot**: {robot_msg}"
            )
            metadata = {"source": f"{date_str}.md"}
            if user_id:
                metadata["user_id"] = user_id
            chunk = Document(text=text, metadata=metadata)
            await reader.index.add_chunk(chunk)
        except Exception as e:
            logging.warning(f"Memory: write-through index failed: {e}")

File: internal/memory/writer.go
package memory

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"go.uber.org/zap"
)

const maxSyncPhotos = 5

// Writer appends interactions to daily markdown files and manages user profiles.
type Writer struct {
	memoryRoot         string
	dailyDir           string
	usersDir           string
	galleryDir         string
	log                *zap.Logger
	visitedThisSession map[string]struct{}
}

// NewWriter creates a Writer, ensuring the required directories exist.
func NewWriter(memoryRoot string, log *zap.Logger) (*Writer, error) {
	dailyDir := filepath.Join(memoryRoot, "daily")
	usersDir := filepath.Join(memoryRoot, "users")

	if err := os.MkdirAll(dailyDir, 0o755); err != nil {
		return nil, fmt.Errorf("memory writer: create daily dir: %w", err)
	}
	if err := os.MkdirAll(usersDir, 0o755); err != nil {
		return nil, fmt.Errorf("memory writer: create users dir: %w", err)
	}

	var galleryDir string
	candidate := filepath.Join(filepath.Dir(memoryRoot), "gallery")
	if info, err := os.Stat(candidate); err == nil && info.IsDir() {
		galleryDir = candidate
		log.Info("memory: gallery detected for photo sync", zap.String("path", galleryDir))
	}

	return &Writer{
		memoryRoot:         memoryRoot,
		dailyDir:           dailyDir,
		usersDir:           usersDir,
		galleryDir:         galleryDir,
		log:                log,
		visitedThisSession: make(map[string]struct{}),
	}, nil
}

// AppendInteraction writes a user message to today's daily log.
func (w *Writer) AppendInteraction(userMsg, uuid, name string) {
	if strings.TrimSpace(userMsg) == "" {
		return
	}

	if uuid != "" {
		w.ensureUserDir(uuid, name)
		w.updateUserProfile(uuid, name)
	}

	dailyPath := w.dailyPath()
	ts := time.Now().Format("15:04:05")

	tag := "unknown"
	if uuid != "" {
		tag = uuid
	}

	entry := fmt.Sprintf("\n## %s\n[User: %s]\n- **User**: %s\n", ts, tag, strings.TrimSpace(userMsg))

	f, err := os.OpenFile(dailyPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		w.log.Error("memory: failed to write interaction", zap.Error(err))
		return
	}
	defer f.Close()

	if _, err := f.WriteString(entry); err != nil {
		w.log.Error("memory: failed to write interaction", zap.Error(err))
	}
}

// AppendToIndex embeds and inserts a new user message into the given index.
func (w *Writer) AppendToIndex(ctx context.Context, idx *MemoryIndex, userMsg, uuid string) {
	if strings.TrimSpace(userMsg) == "" || idx == nil {
		return
	}

	dateStr := time.Now().Format("2006-01-02")
	ts := time.Now().Format("15:04:05")

	tag := "unknown"
	if uuid != "" {
		tag = uuid
	}
	userTag := fmt.Sprintf("[User: %s]\n", tag)

	text := fmt.Sprintf("[Date: %s]\n## %s\n%s- **User**: %s",
		dateStr, ts, userTag, strings.TrimSpace(userMsg))

	meta := map[string]string{"source": dateStr + ".md", "user_id": uuid}

	if _, err := idx.AddChunk(ctx, MemoryEntry{Text: text, Metadata: meta}); err != nil {
		w.log.Warn("memory: write-through index failed", zap.Error(err))
	}
}

func (w *Writer) dailyPath() string {
	today := time.Now().Format("2006-01-02")
	return filepath.Join(w.dailyDir, today+".md")
}

func (w *Writer) ensureUserDir(uuid, name string) {
	userDir := filepath.Join(w.usersDir, uuid)
	_ = os.MkdirAll(userDir, 0o755)

	profilePath := filepath.Join(userDir, "profile.json")
	if _, err := os.Stat(profilePath); os.IsNotExist(err) {
		now := time.Now().Format(time.RFC3339)
		names := []string{}
		if name != "" {
			names = []string{name}
		}
		profile := map[string]any{
			"uuid":              uuid,
			"names":             names,
			"first_seen":        now,
			"last_seen":         now,
			"interaction_count": 0,
			"visit_count":       0,
		}
		data, _ := json.MarshalIndent(profile, "", "  ")
		_ = os.WriteFile(profilePath, data, 0o644)
		w.log.Info("memory: created user profile", zap.String("uuid", uuid), zap.Strings("names", names))
	}

	factsPath := filepath.Join(userDir, "facts.json")
	if _, err := os.Stat(factsPath); os.IsNotExist(err) {
		facts := map[string]any{
			"user_id": uuid,
			"facts":   []any{},
		}
		data, _ := json.MarshalIndent(facts, "", "  ")
		_ = os.WriteFile(factsPath, data, 0o644)
	}

	w.syncPhotos(uuid)
}

// syncPhotos copies the latest maxSyncPhotos aligned photos from the gallery
func (w *Writer) syncPhotos(uuid string) {
	if w.galleryDir == "" {
		return
	}

	srcDir := filepath.Join(w.galleryDir, uuid, "aligned")
	entries, err := os.ReadDir(srcDir)
	if err != nil {
		return
	}

	var jpgs []string
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(strings.ToLower(e.Name()), ".jpg") {
			jpgs = append(jpgs, e.Name())
		}
	}
	if len(jpgs) == 0 {
		return
	}
	sort.Strings(jpgs)

	if len(jpgs) > maxSyncPhotos {
		jpgs = jpgs[len(jpgs)-maxSyncPhotos:]
	}
	wanted := make(map[string]struct{}, len(jpgs))
	for _, name := range jpgs {
		wanted[name] = struct{}{}
	}

	dstDir := filepath.Join(w.usersDir, uuid, "photos")
	_ = os.MkdirAll(dstDir, 0o755)

	for _, name := range jpgs {
		dst := filepath.Join(dstDir, name)
		if _, err := os.Stat(dst); err == nil {
			continue
		}
		if err := copyFile(filepath.Join(srcDir, name), dst); err != nil {
			w.log.Warn("memory: photo sync failed", zap.String("file", name), zap.Error(err))
		}
	}

	existing, _ := os.ReadDir(dstDir)
	for _, e := range existing {
		if _, keep := wanted[e.Name()]; !keep {
			_ = os.Remove(filepath.Join(dstDir, e.Name()))
		}
	}
}

func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, in)
	return err
}

func (w *Writer) updateUserProfile(uuid, name string) {
	profilePath := filepath.Join(w.usersDir, uuid, "profile.json")
	raw, err := os.ReadFile(profilePath)
	if err != nil {
		return
	}

	var profile map[string]any
	if err := json.Unmarshal(raw, &profile); err != nil {
		return
	}

	profile["last_seen"] = time.Now().Format(time.RFC3339)
	count, _ := profile["interaction_count"].(float64)
	profile["interaction_count"] = count + 1

	if name != "" {
		names, _ := profile["names"].([]any)
		found := false
		for _, n := range names {
			if s, ok := n.(string); ok && strings.EqualFold(s, name) {
				found = true
				break
			}
		}
		if !found {
			profile["names"] = append(names, name)
		}
	}

	if _, visited := w.visitedThisSession[uuid]; !visited {
		vc, _ := profile["visit_count"].(float64)
		profile["visit_count"] = vc + 1
		w.visitedThisSession[uuid] = struct{}{}
	}

	data, _ := json.MarshalIndent(profile, "", "  ")
	_ = os.WriteFile(profilePath, data, 0o644)
}
